modules/heat_conduction/src/userobjects/SelfShadowSideUserObject.C

```diff
@@ -0,0 +1,412 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "SelfShadowSideUserObject.h"
+#include "RotationMatrix.h"
+
+#include "libmesh/parallel_algebra.h"
+
+registerMooseObject("HeatConductionApp", SelfShadowSideUserObject);
+
+InputParameters
+SelfShadowSideUserObject::validParams()
+{
+  InputParameters params = SideUserObject::validParams();
+  params.addClassDescription("Compute the illumination status for a self shadowing sideset");
+  params.addRequiredRangeCheckedParam<std::vector<PostprocessorName>>(
+      "illumination_flux",
+      "illumination_flux_size>0 && illumination_flux_size<=3",
+      "Radiation direction vector. Each component of the vector can be a constant number or a "
+      "postprocessor name (the latter enables time varying radiation directions - spatial "
+      "variation is not supported)");
+  return params;
+}
+
+SelfShadowSideUserObject::SelfShadowSideUserObject(const InputParameters & parameters)
+  : SideUserObject(parameters),
+    _dim(_mesh.dimension()),
+    _raw_direction(coupledPostprocessors("illumination_flux"))
+{
+  // we should check the coordinate system (i.e. permit only 0,0,+-1 for RZ)
```

Comment:
Are you planning on checking the mesh's coordinate system here?

Reply:
In a follow on PR.